python/lsst/daf/butler/registry/interfaces/_datasets.py

```diff
@@ -0,0 +1,409 @@
+        """
+        raise NotImplementedError()
+
+    datasetType: DatasetType
+    """Dataset type whose records this object manages (`DatasetType`).
+    """
+
+
+class DatasetRecordStorageManager(ABC):
+    """An interface that manages the tables that describe datasets in a
+    `RegistryLayer`.
+
+    `DatasetRecordStorageManager` primarily serves as a container and factory
+    for `DatasetRecordStorage` instances, which each provide access to the
+    records for a different `DatasetType`.
+    """
+
+    @classmethod
+    @abstractmethod
+    def initialize(cls, db: Database, context: StaticTablesContext, *, collections: CollectionManager,
+                   universe: DimensionUniverse) -> DatasetRecordStorageManager:
+        """Construct an instance of the manager.
+
+        Parameters
+        ----------
+        db : `Database`
+            Interface to the underlying database engine and namespace.
+        context : `StaticTablesContext`
+            Context object obtained from `Database.declareStaticTables`; used
+            to declare any tables that should always be present.
+        collections: `CollectionManager`
+            Manager object for the collections in this `Registry`.
+        universe : `DimensionUniverse`
+            Universe graph containing all dimensions known to this `Registry`.
+
+        Returns
+        -------
+        manager : `DatasetRecordStorageManager`
+            An instance of a concrete `DatasetRecordStorageManager` subclass.
+        """
+        raise NotImplementedError()
+
+    @classmethod
+    @abstractmethod
+    def addDatasetForeignKey(cls, tableSpec: ddl.TableSpec, *, name: str = "dataset",
+                             constraint: bool = True, onDelete: Optional[str] = None,
+                             **kwargs) -> ddl.FieldSpec:
+        """Add a foreign key (field and constraint) referencing the dataset
+        table.
+
+        Parameters
+        ----------
+        tableSpec : `ddl.TableSpec`
+            Specification for the table that should reference the dataset
+            table.  Will be modified in place.
+        name: `str`, optional
+            A name to use for the prefix of the new field; the full name is
+            ``{name}_id``.
+        onDelete: `str`, optional
+            One of "CASCADE" or "SET NULL", indicating what should happen to
+            the referencing row if the collection row is deleted.  `None`
+            indicates that this should be an integrity error.
+        constraint: `bool`, optional
+            If `False` (`True` is default), add a field that can be joined to
+            the dataset primary key, but do not add a foreign key constraint.
+        **kwargs
+            Additional keyword arguments are forwarded to the `ddl.FieldSpec`
+            constructor (only the ``name`` and ``dtype`` arguments are
+            otherwise provided).
+
+        Returns
+        -------
+        idSpec : `ddl.FieldSpec`
```

Comment:
Good question.  I had originally thought that I wouldn't be able to have the various manager classes encapsulate whether they had a compound primary key, and hence any that did need a compound key (as datasets will on DM-21794) would need to make that a part of the ABC interface (and hence DM-21794 would have to break the ABC interface).  But I think encapsulation by returning a list here (and in the getForeignKeyName methods) is actually possible, and is probably a good idea.  That'd be fairly disruptive, but it'd be better to get that disruption over with quickly. I will still probably defer it because this ticket is otherwise pretty close to merge, but I'll open a new ticket and try to get that done soon.